src/mame/f256/f256.cpp

```diff
@@ -0,0 +1,2002 @@
+	m_via6522_1->readpb_handler().set(FUNC(f256_state::via1_system_portb_r));
+	m_via6522_1->writepa_handler().set(FUNC(f256_state::via1_system_porta_w));
+	m_via6522_1->writepb_handler().set(FUNC(f256_state::via1_system_portb_w));
+    m_via6522_1->ca2_handler().set(FUNC(f256_state::via1_ca2_write));
+    m_via6522_1->cb2_handler().set(FUNC(f256_state::via1_cb2_write));
+    m_via6522_1->irq_handler().set(FUNC(f256_state::via1_interrupt));
+
+    SN76489(config, m_sn0, MUSIC_CLOCK / 4);
+    SN76489(config, m_sn1, MUSIC_CLOCK / 4);
+    YMF262(config, m_opl3, MUSIC_CLOCK);
+    MOS6581(config, m_sid0, MUSIC_CLOCK/14);
+    MOS6581(config, m_sid1, MUSIC_CLOCK/14);
+
+    SPEAKER(config, "lspeaker").front_left();
+    SPEAKER(config, "rspeaker").front_right();
+    // Mix PSG
+    m_sn0->add_route(ALL_OUTPUTS, "lspeaker", 0.5);
+    m_sn0->add_route(ALL_OUTPUTS, "rspeaker", 0.5);
+    m_sn1->add_route(ALL_OUTPUTS, "lspeaker", 0.5);
+    m_sn1->add_route(ALL_OUTPUTS, "rspeaker", 0.5);
+
+    m_opl3->add_route(0, "lspeaker", 1.0);
+	m_opl3->add_route(1, "rspeaker", 1.0);
+	m_opl3->add_route(2, "lspeaker", 1.0);
+	m_opl3->add_route(3, "rspeaker", 1.0);
+    // The SIDs are very noisy
+    m_sid0->add_route(ALL_OUTPUTS, "lspeaker", 0.5);
+    m_sid0->add_route(ALL_OUTPUTS, "rspeaker", 0.5);
+    m_sid1->add_route(ALL_OUTPUTS, "lspeaker", 0.5);
+    m_sid1->add_route(ALL_OUTPUTS, "rspeaker", 0.5);
+
+    //set interrupt handler for the RTC
+    m_rtc->int_handler().set(FUNC(f256_state::rtc_interrupt_handler));
+
+    // Add an SD card device
+    SPI_SDCARD(config, m_sdcard, 0);
+	m_sdcard->set_prefer_sdhc();
+    m_sdcard->spi_miso_callback().set([this](int state) {
+        //m_in_bit = state;
+        m_in_latch <<= 1;
+        m_in_latch |= state;
+    });
+
+    m_mouse->set_pc_kbdc(m_ps2_keyboard);
+    NS16550(config, m_uart, MASTER_CLOCK);
+
+    // cbm_iec_slot_device::add(config, m_iec, "c1581");
+	// m_iec->srq_callback().set(FUNC(f256_state::iec_srq_w));
+	// m_iec->data_callback().set(FUNC(f256_state::iec_data_w));
+    // m_iec->atn_callback().set(FUNC(f256_state::iec_atn_w));
+    // m_iec->clk_callback().set(FUNC(f256_state::iec_clk_w));
+}
+f256_state::~f256_state()
+{
+}
+/*
+    Memory map
+    $00:0000 - $07:FFFF	RAM
+    $08:0000 - $0F:FFFF	Flash
+    $10:0000 - $13:FFFF	Expansion Memory - NOT IMPLEMENTED
+    $14:0000 - $1F:FFFF	Reserved
+*/
+void f256_state::program_map(address_map &map)
+{
+    // the address range 0:F
+    // 0: LUT Edit $80 allows writing to 8 to F, LUT Select 0 to 3
+    map(0x0000, 0x000F).rw(FUNC(f256_state::lut_r), FUNC(f256_state::lut_w));
+    map(0x0010, 0xFFFF).rw(FUNC(f256_state::mem_r), FUNC(f256_state::mem_w));
+};
+
+void f256_state::data_map(address_map &map)
+{
+    map(0x0000, 0x1FFF).ram().share(IOPAGE0_TAG);
+    map(0x0000, 0x1FFF).ram().share(IOPAGE1_TAG);
+    map(0x0000, 0x1FFF).ram().share(IOPAGE2_TAG);
+    map(0x0000, 0x1FFF).ram().share(IOPAGE3_TAG);
+}
+
```

Comment:
Huh? This will just map `IOPAGE3_TAG` for the `address_space` rule of last one wins, how it's supposed to really work?